File: backend/apps_manager.py
"""Loading/saving/CRUD for msfs_apps.json (the external addon apps list).

Pure logic, no UI - the Api layer translates results into what the frontend
needs (including translated error messages, since this module only returns
error *codes* matching i18n keys like "err_empty"/"err_delay").
"""
import json
import os

APPS_FILE = "msfs_apps.json"


def load_apps():
    if not os.path.exists(APPS_FILE):
        save_apps({})
        return {}

    with open(APPS_FILE, "r", encoding="utf-8") as f:
        loaded = json.load(f)

    apps = {}
    for name, data in loaded.items():
        if isinstance(data, str):
            delay = 150 if name == "REX Atmos" else 0
            launch_mode = "timer" if delay > 0 else "immediate"
            apps[name] = {"path": data, "delay": delay, "admin": False, "launch_mode": launch_mode}
        else:
            data.setdefault("admin", False)
            data.setdefault("launch_mode", "timer" if data.get("delay", 0) > 0 else "immediate")
            apps[name] = data
    save_apps(apps)
    return apps


def save_apps(apps):
    # Temp file + atomic os.replace, same reasoning as config_manager.py's
    # save_config - a plain "w" open truncates the file immediately, so a
    # crash/kill/power loss mid-write would leave the addon list corrupted.
    tmp_path = APPS_FILE + ".tmp"
    with open(tmp_path, "w", encoding="utf-8") as f:
        json.dump(apps, f, indent=4)
    os.replace(tmp_path, APPS_FILE)


def move_app(apps, name, direction):
    """direction: 'up' or 'down'. Returns the (possibly reordered) dict."""
    keys = list(apps.keys())
    if name not in keys:
        return apps
    idx = keys.index(name)
    target = idx - 1 if direction == "up" else idx + 1
    if 0 <= target < len(keys):
        keys[idx], keys[target] = keys[target], keys[idx]
        apps = {k: apps[k] for k in keys}
        save_apps(apps)
    return apps


def validate_app_input(name, path, mode_key, delay_str):
    """Returns an i18n error key (str) on failure, or None if valid."""
    if not name.strip() or not path.strip():
        return "err_empty"
    if mode_key == "timer":
        try:
            delay = int(delay_str)
            if delay <= 0:
                return "err_delay"
        except (TypeError, ValueError):
            return "err_delay"
    return None


def upsert_app(apps, name, path, mode_key, delay_str, is_admin, previous_name=None):
    """Adds a new app or edits an existing one in place (preserving list position on rename)."""
    delay = int(delay_str) if mode_key == "timer" else 0
    new_data = {"path": path.strip(), "delay": delay, "admin": is_admin, "launch_mode": mode_key}
    name = name.strip()

    if previous_name and previous_name != name and previous_name in apps:
        apps = {
            (name if key == previous_name else key): (new_data if key == previous_name else value)
            for key, value in apps.items()
        }
    else:
        apps[name] = new_data

    save_apps(apps)
    return apps


def remove_app(apps, name):
    apps.pop(name, None)
    save_apps(apps)
    return apps


def open_folder(path):
    folder = os.path.dirname(path)
    if os.path.exists(folder):
        try:
            os.startfile(folder)
        except Exception:
            pass

File: dotnet/src/FlightOpsHub.Core/AppsManager.cs
using System.Diagnostics;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/apps_manager.py - loading/saving/CRUD for
/// msfs_apps.json (the external addon apps list). Pure logic + file IO,
/// no UI - error results are i18n *keys* (e.g. "err_empty"), translated by
/// whatever calls this.
/// </summary>
public static class AppsManager
{
    public const string AppsFileName = "msfs_apps.json";

    public static JsonObject LoadApps(string path)
    {
        if (!File.Exists(path))
        {
            var empty = new JsonObject();
            SaveApps(path, empty);
            return empty;
        }

        JsonObject loaded;
        try
        {
            var text = File.ReadAllText(path, Encoding.UTF8);
            loaded = JsonNode.Parse(text) as JsonObject ?? new JsonObject();
        }
        catch
        {
            loaded = new JsonObject();
        }

        var apps = new JsonObject();
        foreach (var (name, data) in loaded)
        {
            if (data is JsonValue stringValue && stringValue.TryGetValue<string>(out var pathString))
            {
                var delay = name == "REX Atmos" ? 150 : 0;
                apps[name] = new JsonObject
                {
                    ["path"] = pathString,
                    ["delay"] = delay,
                    ["admin"] = false,
                    ["launch_mode"] = delay > 0 ? "timer" : "immediate",
                };
            }
            else if (data is JsonObject obj)
            {
                var clone = (JsonObject)obj.DeepClone();
                clone.TryAdd("admin", false);
                var existingDelay = GetInt(clone, "delay") ?? 0;
                clone.TryAdd("launch_mode", existingDelay > 0 ? "timer" : "immediate");
                apps[name] = clone;
            }
        }
        SaveApps(path, apps);
        return apps;
    }

    public static void SaveApps(string path, JsonObject apps)
    {
        // Temp file + atomic move, same reasoning as ConfigManager.SaveConfig.
        var tmpPath = path + ".tmp";
        try
        {
            // Matches Python's json.dump(apps, f, indent=4); the relaxed
            // encoder avoids .NET's default \uXXXX-escaping of '<'/'>'/'&'
            // etc., which Python's json.dump never does (see ConfigManager
            // and DiagnosticsBuilder for the same fix).
            var options = new JsonSerializerOptions
            {
                WriteIndented = true,
                Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
            };
            File.WriteAllText(tmpPath, apps.ToJsonString(options), new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
            File.Move(tmpPath, path, overwrite: true);
        }
        catch
        {
            try { File.Delete(tmpPath); } catch { /* nothing to clean up */ }
            throw;
        }
    }

    /// <summary>direction: "up" or "down". Persists and returns the (possibly reordered) apps.</summary>
    public static JsonObject MoveApp(string path, JsonObject apps, string name, string direction)
    {
        var keys = apps.Select(kv => kv.Key).ToList();
        var idx = keys.IndexOf(name);
        if (idx < 0) return apps;

        var target = direction == "up" ? idx - 1 : idx + 1;
        if (target < 0 || target >= keys.Count) return apps;

        (keys[idx], keys[target]) = (keys[target], keys[idx]);
        var reordered = new JsonObject();
        foreach (var key in keys)
        {
            reordered[key] = apps[key]?.DeepClone();
        }
        SaveApps(path, reordered);
        return reordered;
    }

    /// <summary>Returns an i18n error key on failure, or null if valid.</summary>
    public static string? ValidateAppInput(string name, string path, string? modeKey, string? delayText)
    {
        if (string.IsNullOrWhiteSpace(name) || string.IsNullOrWhiteSpace(path))
        {
            return "err_empty";
        }
        if (modeKey == "timer")
        {
            if (!int.TryParse(delayText, out var delay) || delay <= 0)
            {
                return "err_delay";
            }
        }
        return null;
    }

    /// <summary>Adds a new app or edits an existing one in place (preserving list position on rename).</summary>
    public static JsonObject UpsertApp(
        string path, JsonObject apps, string name, string appPath, string modeKey,
        string? delayText, bool isAdmin, string? previousName)
    {
        var delay = modeKey == "timer" ? int.Parse(delayText!) : 0;
        var newData = new JsonObject
        {
            ["path"] = appPath.Trim(),
            ["delay"] = delay,
            ["admin"] = isAdmin,
            ["launch_mode"] = modeKey,
        };
        name = name.Trim();

        JsonObject result;
        if (!string.IsNullOrEmpty(previousName) && previousName != name && apps.ContainsKey(previousName))
        {
            result = new JsonObject();
            foreach (var (key, value) in apps)
            {
                result[key == previousName ? name : key] = key == previousName ? newData.DeepClone() : value?.DeepClone();
            }
        }
        else
        {
            result = (JsonObject)apps.DeepClone();
            result[name] = newData;
        }

        SaveApps(path, result);
        return result;
    }

    public static JsonObject RemoveApp(string path, JsonObject apps, string name)
    {
        var result = (JsonObject)apps.DeepClone();
        result.Remove(name);
        SaveApps(path, result);
        return result;
    }

    public static void OpenFolder(string appPath)
    {
        var folder = Path.GetDirectoryName(appPath);
        if (string.IsNullOrEmpty(folder) || !Directory.Exists(folder))
        {
            return;
        }
        try
        {
            Process.Start(new ProcessStartInfo(folder) { UseShellExecute = true });
        }
        catch
        {
            // Best-effort, matches Python's bare except here.
        }
    }

    private static int? GetInt(JsonObject data, string key)
    {
        if (data.TryGetPropertyValue(key, out var node) && node is JsonValue value && value.TryGetValue<int>(out var result))
        {
            return result;
        }
        return null;
    }
}
